### PR TITLE
fix(gnovm): Improve Error Message in `evalStaticTypeOf` function

### LINKED ISSUE #1082: cryptic error: `v := f()`, when f returns no values, panics with `evalStaticTypeOf() only supports *CallExpr with 1 result`
Minimal example to reproduce:

```go
package main

func main() {
        n := f()
}

func f() {
        println("hello!")
}
```

The resulting panic is the following:

```
panic: evalStaticTypeOf() only supports *CallExpr with 1 result, got () [recovered]
        panic: main/a.gno:4: evalStaticTypeOf() only supports *CallExpr with 1 result, got ()

goroutine 1 [running]:
github.com/gnolang/gno/gnovm/pkg/gnolang.Preprocess.func2.1()
        /home/howl/oc/gno/gnovm/pkg/gnolang/preprocess.go:173 +0x46d
panic({0xbf0b00, 0xc00007d450})
        /usr/lib/go/src/runtime/panic.go:890 +0x263
github.com/gnolang/gno/gnovm/pkg/gnolang.evalStaticTypeOf({0xf651b0?, 0xc000188240?}, {0xf62ed8?, 0xc000501800?}, {0xf5c408?, 0xc000128a20?})
        /home/howl/oc/gno/gnovm/pkg/gnolang/preprocess.go:2020 +0xc5
github.com/gnolang/gno/gnovm/pkg/gnolang.Preprocess.func2({0xc00025ae00?, 0x2, 0x20}, 0x46, 0xc000254524?, {0xf5aaf0, 0xc000128a80?}, 0x3)
        /home/howl/oc/gno/gnovm/pkg/gnolang/preprocess.go:1554 +0x8688
github.com/gnolang/gno/gnovm/pkg/gnolang.transcribe(0xc000131970, {0xc00025ae00?, 0x2, 0x20}, 0x46, 0xf5b170?, {0xf5aaf0?, 0xc000128a80}, 0xc000131447)
        /home/howl/oc/gno/gnovm/pkg/gnolang/transcribe.go:728 +0x4b63
github.com/gnolang/gno/gnovm/pkg/gnolang.transcribe(0xc000131970, {0xc00025ae00?, 0x1, 0x20}, 0x4d, 0xf5b0a0?, {0xf5b170?, 0xc000501800}, 0xc000131677)
        /home/howl/oc/gno/gnovm/pkg/gnolang/transcribe.go:670 +0x5125
github.com/gnolang/gno/gnovm/pkg/gnolang.transcribe(0xc000131970, {0xc00025ae00?, 0x0, 0x20}, 0x0, 0x200?, {0xf5b0a0?, 0xc0002b3b80}, 0xc0001318a7)
        /home/howl/oc/gno/gnovm/pkg/gnolang/transcribe.go:708 +0x4cf3
github.com/gnolang/gno/gnovm/pkg/gnolang.Transcribe({0xf5b0a0?, 0xc0002b3b80}, 0x7ffe8a9f6e65?)
        /home/howl/oc/gno/gnovm/pkg/gnolang/transcribe.go:133 +0xbe
github.com/gnolang/gno/gnovm/pkg/gnolang.Preprocess({0xf651b0, 0xc000188240}, {0xf632d8, 0xc0000ca580}, {0xf5b0a0?, 0xc0002b3b80?})
        /home/howl/oc/gno/gnovm/pkg/gnolang/preprocess.go:149 +0x277
github.com/gnolang/gno/gnovm/pkg/gnolang.(*Machine).runFiles(0xc0000c4fc0, {0xc0000123e8, 0x1, 0x1})
        /home/howl/oc/gno/gnovm/pkg/gnolang/machine.go:520 +0x34e
github.com/gnolang/gno/gnovm/pkg/gnolang.(*Machine).RunFiles(...)
```

This is not such a weird mistake to do when programming in Go/Gno, so a panic like this with no reference to the incriminated code is hardly helpful to the developer.

_(Part of a series of issues encountered while developing [gnochess](https://github.com/gnolang/gnochess))_

### PR DESCRIPTION
**Pull request by notJoon**
# Description

Closes #1082 

Enhanced the error message in `evalStaticTypeOf` function to match with go's error message and provide more informative feeback.

## Example of Errors

Sample code (taken from linked issue)

```go
package main

func main() {
	n := f()
}

func f() {
	println("hello!")
}
```

**Before**
```plain
panic: evalStaticTypeOf() only supports *CallExpr with 1 result, got ()
```

**After**
```plain
panic: main/a.gno:3:1: f() (no value) used as value
Hint: Ensure the function returns a single value, or use multiple assignment
```

**Go version (go playground)**
```plain
./prog.go:4:2: declared and not used: n
./prog.go:4:7: f() (no value) used as value
```

## Further Works

I think it would be good to fix `getTypeOf` function as well, but I can't think of a case where it would cause panic.

<details><summary>Contributors' checklist...</summary>

- [X] Added new tests, or not needed, or not feasible
- [X] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [X] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [X] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

File: gnovm/pkg/gnolang/preprocess.go [hunk 1]
```diff
@@ -2449,16 +2449,33 @@ func evalStaticTypeOf(store Store, last BlockNode, x Expr) Type {
...
+func getFunctionName(x Expr) string {
+	switch expr := x.(type) {
+	case *CallExpr:
+		if id, ok := expr.Func.(*NameExpr); ok {
+			return string(id.Name)
 		}
-	} else {
-		return t
 	}
+	return "<unknown function>"
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2478]: should this happen?

File: gnovm/pkg/gnolang/preprocess.go [hunk 2]
```diff
@@ -2449,16 +2449,33 @@ func evalStaticTypeOf(store Store, last BlockNode, x Expr) Type {
...
+		return tt.Elts[0]
+	}
+	return t
+}
+
+// getFunctionName attempts to extract the function name from the expression
+func getFunctionName(x Expr) string {
+	switch expr := x.(type) {
+	case *CallExpr:
+		if id, ok := expr.Func.(*NameExpr); ok {
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2474]: another name other than `id`?

File: gnovm/pkg/gnolang/preprocess.go [hunk 3]
```diff
@@ -2449,16 +2449,33 @@ func evalStaticTypeOf(store Store, last BlockNode, x Expr) Type {
...
+			valueCount := len(tt.Elts)
+
+			var msg string
+			if valueCount == 0 {
+				msg = fmt.Sprintf("%s: %s() (no value) used as value", loc, funcName)
+			} else {
+				msg = fmt.Sprintf("%s: %s() (%d values) used as single value", loc, funcName, valueCount)
+			}
+
+			panic(fmt.Errorf("%s\nHint: Ensure the function returns a single value, or use multiple assignment", msg))
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:2463]: This hint may not be very effective if it corresponds to one of the messages above, such as:

panic: main/func.gno:3:1: f() (no value) used as value
Hint: Ensure the function returns a single value, or use multiple assignment.

I mean in this case, "or use multiple assignment" is superfluous。

**Review by ltzmaxwell — CHANGES_REQUESTED** [review]

File: gnovm/pkg/gnolang/preprocess.go (hunk 3 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:2463]: Or not, this check should be checked ahead of `evalStaticTypeOf` somewhere?

Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:2463]: well I think this part feels odd to me(the original logic in `evalStaticTypeOf`:

```go
if len(tt.Elts) != 1 {
			panic(fmt.Sprintf(
				"evalStaticTypeOf() only supports *CallExpr with 1 result, got %s",
				tt.String(),
			))
```

at the very least for case like:
```go
n := f2()

func f2() (string, int) {
	return "hey", 0
}
```
the check can happened in *AssignStmt, to identify a mismatch on LHS and RHS.

Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:2463]: Writing so much just for the sake of discussion.

File: gnovm/pkg/gnolang/preprocess.go (hunk 1 above)
Comment (notJoon) [reply gnovm/pkg/gnolang/preprocess.go:2478]: This shouldn't happend. I added it just in case.

File: gnovm/pkg/gnolang/preprocess.go [hunk 4]
```diff
@@ -2449,16 +2449,33 @@ func evalStaticTypeOf(store Store, last BlockNode, x Expr) Type {
...
+		}
+		return tt.Elts[0]
+	}
+	return t
+}
+
+// getFunctionName attempts to extract the function name from the expression
+func getFunctionName(x Expr) string {
+	switch expr := x.(type) {
```
Comment (deelawn) [inline gnovm/pkg/gnolang/preprocess.go:2472]: This is a nitpick, but it is a bit strange to have a switch with only one case

File: gnovm/pkg/gnolang/preprocess.go [hunk 5]
```diff
@@ -2449,16 +2449,33 @@ func evalStaticTypeOf(store Store, last BlockNode, x Expr) Type {
...
-			))
-		} else {
-			return tt.Elts[0]
+			loc := last.GetLocation()
+			funcName := getFunctionName(x)
+			valueCount := len(tt.Elts)
+
+			var msg string
+			if valueCount == 0 {
+				msg = fmt.Sprintf("%s: %s() (no value) used as value", loc, funcName)
```
Comment (deelawn) [inline gnovm/pkg/gnolang/preprocess.go:2458]: Maybe the message could be improved here. "no value" doesn't seem as clear as saying something like "function does not return a value".

**Review by deelawn — APPROVED** [review]
A few small suggestions, but looks good to me overall.

**Comment by zivkovicmilos** [comment]
@notJoon 

Did you get a chance to see @ltzmaxwell's comments? 🙏

File: gnovm/pkg/gnolang/preprocess.go (hunk 3 above)
Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go:2463]: Yes @ltzmaxwell I think this check should be done in AssignStmt and ValueDecl in trans_enter

**Review by petar-dambovaliev — APPROVED** [review]

**Comment by ltzmaxwell** [comment]
what is the status of this one?

File: gnovm/pkg/gnolang/preprocess.go (hunk 2 above)
Comment (notJoon) [reply gnovm/pkg/gnolang/preprocess.go]: updated
https://github.com/gnolang/gno/pull/2695/commits/e948f68136acd2d1a05b35fefaa37796dafa65e1

**Comment by notJoon** [comment]
> what is the status of this one?

@ltzmaxwell dust off now. sorry for the delay 🙏

**Comment by Kouteki** [comment]
Closing this PR in favor of https://github.com/gnolang/gno/pull/3049 to avoid duplicate work

**Pull request closed without merging**